pyannote/audio/pipelines/speaker_diarization.py

```diff
@@ -530,20 +538,35 @@ def apply(
         )
         diarization.uri = file["uri"]
 
-        # when reference is available, use it to map hypothesized speakers
-        # to reference speakers (this makes later error analysis easier
-        # but does not modify the actual output of the diarization pipeline)
         if "annotation" in file and file["annotation"]:
-            return self.optimal_mapping(file["annotation"], diarization)
+            # when reference is available, use it to map hypothesized speakers
+            # to reference speakers (this makes later error analysis easier
+            # but does not modify the actual output of the diarization pipeline)
+            label_map = self.optimal_mapping(
+                file["annotation"], diarization, mapping_only=True
+            )
 
-        # when reference is not available, rename hypothesized speakers
-        # to human-readable SPEAKER_00, SPEAKER_01, ...
-        return diarization.rename_labels(
-            {
+        else:
+            # when reference is not available, rename hypothesized speakers
+            # to human-readable SPEAKER_00, SPEAKER_01, ...
+            label_map = {
                 label: expected_label
                 for label, expected_label in zip(diarization.labels(), self.classes())
             }
-        )
+
+        diarization = diarization.rename_labels(label_map)
+
+        if not return_embeddings:
+            return diarization
+        else:
+            embedding_map = {}
+            for k, label in label_map.items():
+                if k < centroids.shape[0]:
+                    embedding_map[label] = centroids[k]
```

Comment:
Can you please clarify what this does and why we need to check that `k < len(centroids)`.

This part calls for a comment in the code explaining exactly the assumption made regarding the type (`int`) and values (`0` to `k-1`?) of speaker labels returned by `self.to_annotation`.

Reply:
The number of final speaker labels can be larger than the number of returned centroids due to the issue with the speaker counting. This issue is addressed in the next PR #1351. Before those changes are applied, we have to check that initial numeric speaker labels assigned after `to_annotation` do not exceed actual number of centroids found. This is the main reason I initially decided to merge those changes together, since the situation when the number of returned centroids is less than the number of speaker labels in annotation is semantically incorrect.